core/amber/src/main/python/core/python_worker.py

```diff
@@ -31,12 +57,66 @@ def run(self) -> None:
         )
         main_loop_thread = Thread(target=self._main_loop.run, name="main_loop_thread")
 
+        stop_event = Event()
+        heartbeat_thread = Thread(
+            target=self.heartbeat,
+            name="heartbeat_thread",
+            args=(5.0, stop_event, network_sender_thread, main_loop_thread),
+        )
+
         network_sender_thread.start()
         main_loop_thread.start()
+        heartbeat_thread.start()
         main_loop_thread.join()
         network_sender_thread.join()
+        with self.condition:
+            self.condition.notify()
+
+        heartbeat_thread.join()
 
     @overrides
     def stop(self):
         self._main_loop.stop()
         self._network_sender.stop()
+        self_clean_child_process()
+
+    def heartbeat(self, interval, stop_event, network_sender_thread, main_loop_thread):
+        while network_sender_thread.is_alive() and main_loop_thread.is_alive():
+            try:
+                socket.create_connection(
+                    (self.parsed_server_host, self.parsed_server_port), timeout=1
+                )
+            except Exception as e:
+                logger.warning(f"Server is down with exception: {e}")
+                stop_event.set()
+                break
+            with self.condition:
+                self.condition.wait(interval)
+        try:
+            socket.create_connection(
+                (self.parsed_server_host, self.parsed_server_port), timeout=1
+            )
+        except Exception as e:
+            logger.warning(f"Server is down with exception: {e}")
+            stop_event.set()
```

Comment:
it seems that the `stop_event` is set and assessed in the same thread. If so maybe it doesn't need to be an event. Can we use a boolean value instead?